File: desktop/src/sdn-updater/carrier.js
// SDN update carrier parsing. The WASM carrier is a storage envelope only:
// the bundle bytes are read out of a custom section without ever compiling,
// instantiating, or executing the module. This is the Node twin of
// sdn-server/internal/update/carrier.go and must keep the section name and
// wire format identical.

const WASM_BUNDLE_SECTION_NAME = 'sdn.update.bundle.v1'
const WASM_MAGIC = Buffer.from([0x00, 0x61, 0x73, 0x6d])
const WASM_HEADER_LENGTH = 8
const CUSTOM_SECTION_ID = 0

function normalizeBytes (value) {
  if (Buffer.isBuffer(value)) {
    return value
  }
  if (value instanceof Uint8Array) {
    return Buffer.from(value.buffer, value.byteOffset, value.byteLength)
  }
  if (value instanceof ArrayBuffer) {
    return Buffer.from(value)
  }

  throw new Error('update carrier is not a wasm module')
}

// Reads an unsigned LEB128 varuint32 at offset. Returns { value, length }.
function readVaruint32 (bytes, offset) {
  let value = 0
  let length = 0

  while (true) {
    if (offset + length >= bytes.length || length >= 5) {
      throw new Error('malformed update carrier: invalid varuint32')
    }
    const byte = bytes[offset + length]
    value += (byte & 0x7f) * 2 ** (7 * length)
    length++
    if ((byte & 0x80) === 0) {
      break
    }
  }

  if (value > 0xffffffff) {
    throw new Error('malformed update carrier: invalid varuint32')
  }

  return { value, length }
}

// Walks the WASM module sections without instantiating the module and returns
// the payload bytes of the SDN bundle custom section.
function extractBundleBytes (wasmBytes) {
  const bytes = normalizeBytes(wasmBytes)

  if (bytes.length < WASM_HEADER_LENGTH || !bytes.subarray(0, 4).equals(WASM_MAGIC)) {
    throw new Error('update carrier is not a wasm module')
  }

  let offset = WASM_HEADER_LENGTH
  while (offset < bytes.length) {
    const sectionId = bytes[offset]
    offset++

    const sectionSize = readVaruint32(bytes, offset)
    offset += sectionSize.length

    if (offset + sectionSize.value > bytes.length) {
      throw new Error('malformed update carrier: truncated section')
    }
    const payload = bytes.subarray(offset, offset + sectionSize.value)
    offset += sectionSize.value

    if (sectionId !== CUSTOM_SECTION_ID) {
      continue
    }

    const nameLength = readVaruint32(payload, 0)
    if (nameLength.length + nameLength.value > payload.length) {
      throw new Error('malformed update carrier: truncated section name')
    }
    const name = payload.subarray(nameLength.length, nameLength.length + nameLength.value).toString('utf8')
    if (name === WASM_BUNDLE_SECTION_NAME) {
      return payload.subarray(nameLength.length + nameLength.value)
    }
  }

  throw new Error('update carrier does not contain an SDN bundle section')
}

module.exports = {
  WASM_BUNDLE_SECTION_NAME,
  extractBundleBytes
}

File: sdn-server/internal/update/carrier.go
package update

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
)

// BundleSectionName is the WASM custom section that carries the compressed
// update bundle inside the inert update.wasm carrier. The carrier is a
// storage envelope only; it is never instantiated or executed.
const BundleSectionName = "sdn.update.bundle.v1"

var wasmMagic = []byte{0x00, 0x61, 0x73, 0x6d}

// ExtractBundleFromCarrier walks the WASM module's sections without
// instantiating it and returns the payload of the SDN bundle custom section.
func ExtractBundleFromCarrier(wasmBytes []byte) ([]byte, error) {
	if len(wasmBytes) < 8 || !bytes.Equal(wasmBytes[:4], wasmMagic) {
		return nil, errors.New("update carrier is not a wasm module")
	}
	offset := 8
	for offset < len(wasmBytes) {
		sectionID := wasmBytes[offset]
		offset++
		sectionSize, n, err := readVaruint32(wasmBytes[offset:])
		if err != nil {
			return nil, fmt.Errorf("malformed update carrier: %w", err)
		}
		offset += n
		if offset+int(sectionSize) > len(wasmBytes) {
			return nil, errors.New("malformed update carrier: truncated section")
		}
		payload := wasmBytes[offset : offset+int(sectionSize)]
		offset += int(sectionSize)
		if sectionID != 0 {
			continue
		}
		nameLen, n, err := readVaruint32(payload)
		if err != nil {
			return nil, fmt.Errorf("malformed update carrier: %w", err)
		}
		if n+int(nameLen) > len(payload) {
			return nil, errors.New("malformed update carrier: truncated section name")
		}
		name := string(payload[n : n+int(nameLen)])
		if name == BundleSectionName {
			return payload[n+int(nameLen):], nil
		}
	}
	return nil, errors.New("update carrier does not contain an SDN bundle section")
}

// BuildCarrier wraps bundle bytes in a minimal valid WASM module containing
// only the SDN bundle custom section. Used by release tooling and tests; the
// canonical Node implementation lives in deployment/release.
func BuildCarrier(bundleBytes []byte) []byte {
	name := []byte(BundleSectionName)
	payload := append(appendVaruint32(nil, uint32(len(name))), name...)
	payload = append(payload, bundleBytes...)

	module := append([]byte{}, wasmMagic...)
	module = append(module, 0x01, 0x00, 0x00, 0x00)
	module = append(module, 0x00)
	module = appendVaruint32(module, uint32(len(payload)))
	return append(module, payload...)
}

func readVaruint32(data []byte) (uint32, int, error) {
	value, n := binary.Uvarint(data)
	if n <= 0 || value > 0xffffffff {
		return 0, 0, errors.New("invalid varuint32")
	}
	return uint32(value), n, nil
}

func appendVaruint32(dst []byte, value uint32) []byte {
	return binary.AppendUvarint(dst, uint64(value))
}
